src/sentry/integrations/vsts/integration.py

```diff
@@ -0,0 +1,163 @@
+            if project['id'] == project_id:
+                return project
+        return None
+
+
+class ProjectForm(forms.Form):
+    def __init__(self, projects, *args, **kwargs):
+        super(ProjectForm, self).__init__(*args, **kwargs)
+        self.fields['project'] = forms.ChoiceField(
+            choices=[(project['id'], project['name']) for project in projects],
+            label='Project',
+            help_text='Enter the Visual Studio Team Services project name that you wish to use as a default for new work items'
+        )
+
+
+class VSTSIntegration(Integration):
+    key = 'vsts'
+    name = 'Visual Studio Team Services'
+    metadata = metadata
+    domain = '.visualstudio.com'
+    api_version = '4.1'
+
+    setup_dialog_config = {
+        'width': 600,
+        'height': 800,
+    }
+
+    identity_oauth_scopes = frozenset([
+        'vso.build_execute',
+        'vso.code_full',
+        'vso.codesearch',
+        'vso.connected_server',
+        'vso.dashboards_manage',
+        'vso.entitlements',
+        'vso.extension.data_write',
+        'vso.extension_manage',
+        'vso.gallery_manage',
+        'vso.graph_manage',
+        'vso.identity_manage',
+        'vso.loadtest',
+        'vso.machinegroup_manage',
+        'vso.memberentitlementmanagement_write',
+        'vso.notification_diagnostics',
+        'vso.notification_manage',
+        'vso.packaging_manage',
+        'vso.profile_write',
+        'vso.project_manage',
+        'vso.release_manage',
+        'vso.security_manage',
+        'vso.serviceendpoint_manage',
+        'vso.symbols_manage',
+        'vso.taskgroups_manage',
+        'vso.test_write',
+        'vso.wiki_write',
+        'vso.work_full',
+        'vso.workitemsearch',
+    ])
+
+    def get_pipeline_views(self):
+        identity_pipeline_config = {
+            'oauth_scopes': self.identity_oauth_scopes,
+            'redirect_url': absolute_uri('/extensions/vsts/setup/'),
+        }
+
+        identity_pipeline_view = NestedPipelineView(
+            bind_key='identity',
+            provider_key='vsts',
+            pipeline_cls=IdentityProviderPipeline,
+            config=identity_pipeline_config,
+        )
+
+        return [
+            identity_pipeline_view,
+            ProjectConfigView(),
+        ]
+
+    def build_integration(self, state):
+        data = state['identity']['data']
+        account = state['identity']['account']
+        access_token = data['access_token']
+        instance = state['identity']['instance']
+        project = state['project']
+
+        scopes = sorted(self.identity_oauth_scopes)
+        return {
+            'name': project['name'],
+            'external_id': project['id'],
+            'metadata': {
+                'domain_name': instance,
+                'scopes': scopes,
+                # icon doesn't appear to be possible
+            },
+            'user_identity': {
+                'access_token': access_token,
```

Comment:
Access token actually goes in the `data` object, sorry should have made that more clear. Check out the `get_oauth_data` method used in the slack identity provider to get the oauth data key (Would be nice if we could reuse this, but would need some refactoring).

https://github.com/getsentry/sentry/blob/c62a7b251ee9fc840a81621374ba34bf46a7ceb5/src/sentry/identity/oauth2.py#L56-L66

Reply:
This is definitely a bit different from the Slack integration, since fundamentally the access token that comes back from slack is for an entire workspace, where as this token is for a users account.